examples/example3/README.md

```diff
@@ -0,0 +1,40 @@
+## Example 3 – Using virtual resources and collection in a master-agent set-up.
+
+
+This is an (almost) complete example for a master-agent set-up using virtual resources in Puppet.
```

Comment:
Specifically I'm referring to `icinga2::object::service` as an exported resource, as this is the resource type that has the most room for growth in monitoring infrastructure.  Essentially, you just want to get data from a node in your infrastructure to the monitoring server so that icinga knows what to monitor.  If you are already exporting a `host` object which has a pile of vars that you care about, then you should be able to use that data to avoid exporting resources for the bulk of the `service` checks.  The problem specifically is that the number of exports compounds as you add more nodes to the infrastructure for base system checks, and can lead to performance issues and cause people shy away from exported resources entirely.

From my use of icinga2 thus far, the apply rules here are really a saving grace, since icinga2 has a language built in to apply service checks to a host using only data that icinga2 knows about.

That said, I do have service checks that are exported, but I try to keep them relatively low.  Also, I've been using Puppet for a long time, and so perhaps the issues that I've seen many moons ago are less present than they are today.

Another option is to use the puppetdbquery module to ask puppetdb direclty for information about which nodes have which classes and then use the data to determine which checks to apply to which nodes.

I only mention this because I'd like to avoid recommendations that can lead to catalog compile times increasing without remorse.  And further, I'd like to give recommendations to folks, as has been done here, for how to be successful in icing2 and Puppet.  That was really the only reason for my initial comment.

I've only rolled icinga2 with puppet in two infrastructures, and on the old module with hopes to migrate them to the new module.  I'm happy to provide opinionated code snippets if thats helpful.  I plan to be using puppet and icinga2 for quite a while, though I still need to migrate to this module.

If anyone has more modern data on this subject, I'd love to hear about it.